src/algebra/regular.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+
+/-!
+# Regular elements
+
+We introduce left-regular, right-regular and regular elements.
+
+By definition, a regular element in a commutative ring is a non-zero divisor.
+Lemma `is_regular_of_integral_domain` shows that every non-zero element of an integral domain
+is regular.
+
+The final goal is to develop part of the API to prove, eventually, results about non-zero-divisors.
+-/
+variables {R : Type*}
+
+/-- A left-regular element is an element `c` such that multiplication on the left by `c`
+is injective on the left. -/
+def is_left_regular [has_mul R] (c : R) := function.injective ((*) c)
+
+/-- A right-regular element is an element `c` such that multiplication on the right by `c`
+is injective on the left. -/
+def is_right_regular [has_mul R] (c : R) := function.injective (* c)
+
+/-- A regular element is an element `c` such that multiplication by `c` both on the left and
+on the right is injective. -/
+structure is_regular [has_mul R] (c : R) : Prop :=
+(left : is_left_regular c)
+(right : is_right_regular c)
+
+namespace is_regular
+
+variables [monoid R]
+
+/-- An element admitting a right inverse is right-regular. -/
+lemma right_of_mul_eq_one {a ai : R} (h : a * ai = 1) : is_right_regular a :=
+begin
+  intros b c bc,
+  rw [← mul_one b, ← mul_one c, ← h, ← mul_assoc, ← mul_assoc],
+  exact congr_fun (congr_arg has_mul.mul bc) ai,
+end
+
+/-- An element admitting a left inverse is left-regular. -/
+lemma left_of_mul_eq_one {a ai : R} (h : ai * a = 1) : is_left_regular a :=
+begin
+  intros b c bc,
+  rw [← one_mul b, ← one_mul c, ← h, mul_assoc, mul_assoc],
+  exact congr_arg (has_mul.mul ai) bc,
+end
+
+/-- A left and right regular element is regular. -/
+lemma of_left_right {a : R} (hl : is_left_regular a) (hr : is_right_regular a) : is_regular a :=
+⟨hl, hr⟩
+
+/-- An element admitting a left and a right inverse is regular. -/
+lemma of_mul_eq_one_mul_eq_one {a ai : R} (hr : a * ai = 1) (hl : ai * a = 1) : is_regular a :=
+of_left_right (left_of_mul_eq_one hl) (right_of_mul_eq_one hr)
+
+/--  Elements of a left cancel semigroup are left regular.
+
+Funny how left and right change positions. -/
```

Comment:
This is a comment about the proof, not about the API - a viewer of docgen will have no clue what this remark refers to.

If you want to keep the comment, make it a non-doc comment.

Reply:
I omitted the comment, leaving only the doc-string